### PR TITLE
Aes encryption with password

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: client/src/main/scala/fluence/client/AuthorizedClient.scala [hunk 1]
```diff
@@ -27,12 +27,16 @@ import scala.language.higherKinds
...
+ * @param keyPair A pair of keys with a public key that will be used as an address for dataset ids and contracts
+ * @param key Key for encryption of keys and values in data storage
  */
-case class AuthorizedClient(kp: KeyPair)
+case class AuthorizedClient[T](keyPair: KeyPair, key: T)
 
 object AuthorizedClient {
-  def generateNew[F[_] : Monad](signAlgo: SignAlgo): EitherT[F, CryptoErr, AuthorizedClient] = {
-    signAlgo.generateKeyPair[F]().map(AuthorizedClient.apply)
+  //it is better to save the password as an char array than as a string
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/AuthorizedClient.scala]: why?)

File: client/src/main/scala/fluence/client/ClientApp.scala [hunk 2]
```diff
@@ -76,16 +77,21 @@ object ClientApp extends App with slogging.LazyLogging {
...
+      val io = for {
+        res ← task.toIO
         _ = {
           logger.info("You can put or get data from remote node.")
           logger.info("Examples: ")
           logger.info("put \"some key\" \"value to put\"")
           logger.info("get \"some key\"")
         }
-      } yield (keyPair, fluenceClient)
+        pass ← readPassword()
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/ClientApp.scala]: Is there is no variant to not use cryptography?

File: client/src/main/scala/fluence/client/ClientApp.scala [hunk 3]
```diff
@@ -100,8 +106,16 @@ object ClientApp extends App with slogging.LazyLogging {
     //Try --help for more information.
   }
 
-  def handleCmds(fluenceClient: FluenceClient, ac: AuthorizedClient): IO[Unit] = {
-    val readLine = IO(lineReader.readLine("fluence< "))
+  def readLine = IO(lineReader.readLine("fluence< "))
+
+  private def readPassword() = {
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/ClientApp.scala]: Nice to have cryptography as option, not as requirement

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 4]
```diff
@@ -83,27 +85,29 @@ class FluenceClient(
...
-    addDataset(ac, storageRpc(contact), NoOpCrypt.forString, NoOpCrypt.forString, clientState, storageHasher)
+  private def addEncryptedDataset(ac: AuthorizedClient[Password], contact: Contact, clientState: Option[ClientState]): Task[ClientDatasetStorage[String, String]] = {
+    AesConfigParser.readAesConfigOrGetDefault(config).flatMap { aesConfig ⇒
+      addDataset(
+        ac,
+        storageRpc(contact),
+        AesCrypt.forString(ac.key, withIV = false, config = aesConfig),
+        AesCrypt.forString(ac.key, withIV = false, config = aesConfig),
+        clientState,
+        storageHasher)
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]:
```
   storageHasher
)
```

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 5]
```diff
@@ -268,7 +272,8 @@ object FluenceClient extends slogging.LazyLogging {
     contracts: Contracts[Task, BasicContract, Contact],
     storageRpc: Contact ⇒ DatasetStorageRpc[Task],
     signAlgo: SignAlgo = Ecdsa.signAlgo,
-    storageHasher: CryptoHasher[Array[Byte], Array[Byte]]): FluenceClient = {
-    new FluenceClient(kademliaClient, contracts, signAlgo, storageRpc, storageHasher)
+    storageHasher: CryptoHasher[Array[Byte], Array[Byte]],
+    config: Config): FluenceClient = {
+    new FluenceClient(kademliaClient, contracts, signAlgo, storageRpc, storageHasher, config)
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]:
```	
    storageHasher: CryptoHasher[Array[Byte], Array[Byte]],
    config: Config
): FluenceClient = {
     new FluenceClient(kademliaClient, contracts, signAlgo, storageRpc, storageHasher, config)
}

File: crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesConfig.scala [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package fluence.crypto.algorithm
```
Comment (RicoGit) [inline crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesConfig.scala]: header!

File: crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesConfig.scala [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+
+case class AesConfig(
```
Comment (RicoGit) [inline crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesConfig.scala]: Docs needed for class and fields.

File: crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesCrypt.scala [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+ * AES with CBC BC - Advanced Encryption Standard with Cipher Block Chaining
+ * https://ru.wikipedia.org/wiki/Advanced_Encryption_Standard
+ * https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Cipher_Block_Chaining_(CBC)
+ * @param password User entered password
+ * @param withIV Initialization vector to achieve semantic security, a property whereby repeated usage of the scheme
+ *               under the same key does not allow an attacker to infer relationships between segments of the encrypted
+ *               message
+ */
+class AesCrypt[F[_] : Monad, T](password: Array[Char], withIV: Boolean, config: AesConfig)(serializer: T ⇒ F[Array[Byte]], deserializer: Array[Byte] ⇒ F[T])(implicit ME: MonadError[F, Throwable])
+  extends Crypt[F, T, Array[Byte]] with JavaAlgorithm {
```
Comment (RicoGit) [inline crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesCrypt.scala]: Why did you use serializer and deserializer instead Codec?

File: crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesCrypt.scala [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+      dataWithParams ← detachDataAndGetParams(cipherText, password, salt, withIV)
+      decData ← processData(dataWithParams, None, encrypt = false)
+      plain ← EitherT.liftF[F, CryptoErr, T](deserializer(decData))
+    } yield plain
+
+    e.value.flatMap(ME.fromEither)
+  }
+
+  /**
+   * key spec initialization
```
Comment (RicoGit) [inline crypto/jvm/src/main/scala/fluence/crypto/algorithm/AesCrypt.scala]: Why doc starts with small char?

File: node/src/test/scala/fluence/node/ClientNodeIntegrationSpec.scala [hunk 10]
```diff
@@ -364,7 +367,7 @@ class ClientNodeIntegrationSpec extends WordSpec with Matchers with ScalaFutures
     val getKey1Response = datasetStorage.get("key1").taskValue
     getKey1Response shouldBe Some("value1")
     // override value
-    val overrideResponse = datasetStorage.put("key1", "value1-NEW").taskValue
+    val overrideResponse = datasetStorage.put("key1", "value1-NEW").taskValue(Some(timeout(Span(1, Seconds))))
```
Comment (RicoGit) [inline node/src/test/scala/fluence/node/ClientNodeIntegrationSpec.scala]: Why there is 1 second, global timeout is 5 seconds? Do you really need to decrease timeout here?

File: client/src/main/scala/fluence/client/AuthorizedClient.scala (hunk 1 above)
Comment (DieMyst) [reply client/src/main/scala/fluence/client/AuthorizedClient.scala]: https://stackoverflow.com/questions/8881291/why-is-char-preferred-over-string-for-passwords

File: client/src/main/scala/fluence/client/ClientApp.scala (hunk 2 above)
Comment (DieMyst) [reply client/src/main/scala/fluence/client/ClientApp.scala]: we decide to use private key as password for demo, so i will delete this

File: client/src/main/scala/fluence/client/FluenceClient.scala [hunk 11]
```diff
@@ -264,11 +261,13 @@ object FluenceClient extends slogging.LazyLogging {
    * @param signAlgo main algorithm for key verification
    */
   def apply(
-    kademliaClient: Kademlia[Task, Contact],
-    contracts: Contracts[Task, BasicContract, Contact],
-    storageRpc: Contact ⇒ DatasetStorageRpc[Task],
-    signAlgo: SignAlgo = Ecdsa.signAlgo,
-    storageHasher: CryptoHasher[Array[Byte], Array[Byte]]): FluenceClient = {
-    new FluenceClient(kademliaClient, contracts, signAlgo, storageRpc, storageHasher)
+            kademliaClient: Kademlia[Task, Contact],
```
Comment (RicoGit) [inline client/src/main/scala/fluence/client/FluenceClient.scala]: there is so many spaces

**Pull request merged**